src/ripple/app/tx/impl/PayChan.cpp

```diff
@@ -0,0 +1,470 @@
+
+        if (balance < reserve + ctx_.tx[sfAmount])
+            return tecUNFUNDED;
+    }
+
+    auto const dst = ctx_.tx[sfDestination];
+
+    {
+        // Check destination account
+        auto const sled = ctx_.view ().read (keylet::account (dst));
+        if (!sled)
+            return tecNO_DST;
+        if (((*sled)[sfFlags] & lsfRequireDestTag) &&
+            !ctx_.tx[~sfDestinationTag])
+            return tecDST_TAG_NEEDED;
+        if ((*sled)[sfFlags] & lsfDisallowXRP)
+            return tecNO_TARGET;
+    }
+
+    // Create PayChan in ledger
+    auto const slep = std::make_shared<SLE> (
+        keylet::payChan (account, dst, (*sle)[sfSequence] - 1));
+    // Funds held in this channel
+    (*slep)[sfAmount] = ctx_.tx[sfAmount];
+    // Amount channel has already paid
+    (*slep)[sfBalance] = ctx_.tx[sfAmount].zeroed ();
+    (*slep)[sfAccount] = account;
+    (*slep)[sfDestination] = dst;
+    (*slep)[sfSettleDelay] = ctx_.tx[sfSettleDelay];
+    (*slep)[sfPublicKey] = ctx_.tx[sfPublicKey];
+    (*slep)[~sfCancelAfter] = ctx_.tx[~sfCancelAfter];
+    (*slep)[~sfSourceTag] = ctx_.tx[~sfSourceTag];
+    (*slep)[~sfDestinationTag] = ctx_.tx[~sfDestinationTag];
+
+    ctx_.view ().insert (slep);
+
+    // Add PayChan to owner directory
+    {
+        uint64_t page;
+        auto result = dirAdd (ctx_.view (), page, keylet::ownerDir (account),
+            slep->key (), describeOwnerDir (account),
+            ctx_.app.journal ("View"));
+        if (!isTesSuccess (result.first))
+            return result.first;
+        (*slep)[sfOwnerNode] = page;
+    }
+
+    // Deduct owner's balance, increment owner count
+    (*sle)[sfBalance] = (*sle)[sfBalance] - ctx_.tx[sfAmount];
+    (*sle)[sfOwnerCount] = (*sle)[sfOwnerCount] + 1;
+    ctx_.view ().update (sle);
+
+    return tesSUCCESS;
+}
+
+//------------------------------------------------------------------------------
+
+TER
+PayChanFund::preflight (PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled (featurePayChan, ctx.app.config ().features))
+        return temDISABLED;
+
+    auto const ret = preflight1 (ctx);
+    if (!isTesSuccess (ret))
+        return ret;
+
+    if (!isXRP (ctx.tx[sfAmount]) || (ctx.tx[sfAmount] <= beast::zero))
+        return temBAD_AMOUNT;
+
+    return preflight2 (ctx);
+}
+
+TER
+PayChanFund::doApply()
+{
+    Keylet const k (ltPAYCHAN, ctx_.tx[sfPayChannel]);
+    auto const slep = ctx_.view ().peek (k);
+    if (!slep)
+        return tecNO_ENTRY;
+
+    AccountID const src = (*slep)[sfAccount];
+    auto const txAccount = ctx_.tx[sfAccount];
+    auto const expiration = (*slep)[~sfExpiration];
+
+    {
+        auto const cancelAfter = (*slep)[~sfCancelAfter];
+        auto const closeTime =
+            ctx_.view ().info ().parentCloseTime.time_since_epoch ().count ();
+        if ((cancelAfter && closeTime >= *cancelAfter) ||
+            (expiration && closeTime >= *expiration))
+            return closeChannel (
```

Comment:
Addendum: `tesSUCCESS` is the best we can do now because `closeChannel` will transfer any unclaimed funds back to the source,and a `tec` code can't really perform transactions.

The documentation needs to explicitly explain that "fund" really means "fund if not expired, do nothing otherwise" and that a `tesSUCCESS` may mean "funding succeeded!" or "channel expired."